waterlog/src/Components/Map.js

```diff
@@ -1,132 +1,189 @@
-			markers: markers,
-			simpleView: false,
-			segments: segments,
-			zoom: 17
-		};
-	}
+function formatMapData(data) {
+  let markers, segments;
+  let todaysLeaks;
+  if (!data.leaks) {
+    todaysLeaks = [];
+  } else {
+    todaysLeaks = data.leaks.filter(leak => (
+      (new Date(leak.latestTimeStamp)).getDay() === (new Date(Date.now())).getDay() &&
+      (new Date(leak.latestTimeStamp)).getMonth() === (new Date(Date.now())).getMonth() &&
+      (new Date(leak.latestTimeStamp)).getFullYear() === (new Date(Date.now())).getFullYear()
+    ));
+  }
+
+  if (data.segments) {
+    segments = data.segments.map(seg => {
+      const leak = todaysLeaks.find(leak => leak.segmentsId === seg.id);
+      if (leak) {
+        seg.status = (leak.resolvedStatus.toLowerCase() === "unresolved") ? "leak" : "normal";
+      } else {
+        seg.status = 'normal'
+      }
+      return seg;
+    });
+  } else {
+    return [];
+  }
+
+  markers = data.monitors.map(mon => {
+    return { id: mon.id, lat: mon.lat, lon: mon.long, status: mon.status }
+  });
+  return { markers, segments };
+}
+
+function generateMapIcons({ segments, markers }, simpleView) {
+  const defaultColor = simpleView ? backgroundColor : lighterColor;
+  if (!segments || !markers) {
+    return <div></div>
+  }
+  return segments.map((segment) => {
+    const sensorIn = markers.find(marker => (marker.id === segment.senseIDIn));
+    const sensorOut = markers.find(marker => (marker.id === segment.senseIDOut));
 
-	render() {
-    const position = [ this.state.lat, this.state.lng ];
-    let defaultColor = this.state.simpleView ? backgroundColor:lighterColor;
-		return (
-			<div className="map-main-div">
-				<div className="map-tile-div">
-					<Map center={position} zoom={this.state.zoom} zoomControl={false}>
-						{(() => {
-							if (this.state.simpleView)
-								return <TileLayer url="https://{s}.basemaps.cartocdn.com/dark_all/{z}/{x}/{y}{r}.png" />;
-						})()}
-						{this.state.segments.map((segment) => { 
-							let sensorIn = this.state.markers.find((marker) => {
-								if (marker.id === segment.senseIDIn) return marker; else return null;
-							});
-							let sensorOut = this.state.markers.find((marker) => {
-								if (marker.id === segment.senseIDOut) return marker; else return null;
-							});
-							let sensorInColor = defaultColor,
-								sensorOutColor = defaultColor,
-								segmentColor = defaultColor;
-							if (sensorIn.status.toLowerCase() === 'fault') {
-								sensorInColor = errorColor;
-							}
-							if (sensorOut.status.toLowerCase() === 'fault') {
-								sensorOutColor = errorColor;
-							}
-							if (segment.status.toLowerCase() === 'leak') {
-								segmentColor = errorColor;
-							}
+    if (sensorIn === undefined || sensorOut === undefined) return <div></div>;
+    let sensorInColor = defaultColor,
+      sensorOutColor = defaultColor,
+      segmentColor = defaultColor;
+    if (sensorIn.status.toLowerCase() === 'fault') {
```

Comment:
should you test if sensorIn.status exists before checking sensorIn.status.toLowercase?

Reply:
I used this line "if (sensorIn === undefined || sensorOut === undefined) return <div></div>;" to test if the variable is valid.

Reply:
but not sensorIn.status

Reply:
sensorIn might exist yes, but that does not mean sensorIn.status exist

Reply:
Ohh yah, I see. I will update that